### PR TITLE
IAM util refactors 

### PR DESCRIPTION
**Pull request by emilymye**
Precursor to batching IAM requests
- Moved common utils from `resource_google_project_iam_policy.go` to `iam.go`
- Moved `iam.go` unit tests to their own file
- Add flatten/expanders for conversion of list of Binding/AuditConfigs to/from golangSet
- Misc refactors for utils in `iam.go`: use struct{}{} instead of bool for sets, comments

# Release Note for Downstream PRs (will be copied)
```releasenote

```

### CODE REVIEW

File: third_party/terraform/resources/resource_google_project_iam_policy.go [hunk 1]
```diff
@@ -188,121 +184,6 @@ func getProjectIamPolicy(project string, config *Config) (*cloudresourcemanager.
 	return p, nil
 }
 
-func jsonPolicyDiffSuppress(k, old, new string, d *schema.ResourceData) bool {
```
Comment (emilymye) [inline third_party/terraform/resources/resource_google_project_iam_policy.go]: These are now in iam.go

File: third_party/terraform/resources/resource_google_project_iam_policy.go [hunk 2]
```diff
@@ -188,121 +184,6 @@ func getProjectIamPolicy(project string, config *Config) (*cloudresourcemanager.
...
-				if b[i].AuditLogConfigs[x].ExemptedMembers[pos] != exemptedMember {
-					return false
-				}
-			}
-		}
-	}
-	return true
-}
-
-type sortableBindings []*cloudresourcemanager.Binding
```
Comment (emilymye) [inline third_party/terraform/resources/resource_google_project_iam_policy.go]: I removed these sorting functions in factor of converting to a golang set and then comparing

File: third_party/terraform/tests/resource_google_project_iam_policy_test.go [hunk 3]
```diff
@@ -158,152 +156,17 @@ func testAccCheckGoogleProjectIamPolicyExists(projectRes, policyRes, pid string)
...
 		// The audit configs in both policies should be identical
 		if !compareAuditConfigs(projectPolicy.AuditConfigs, policyPolicy.AuditConfigs) {
-			return fmt.Errorf("Project and data source policies do not match: project policy is %+v, data resource policy is  %+v", projectPolicy.AuditConfigs, policyPolicy.AuditConfigs)
+			return fmt.Errorf("Project and data source policies do not match: project policy is %+v, data resource policy is  %+v", debugPrintAuditConfigs(projectPolicy.AuditConfigs), debugPrintAuditConfigs(policyPolicy.AuditConfigs))
 		}
 		return nil
 	}
 }
 
-func TestIamOverwriteBinding(t *testing.T) {
```
Comment (emilymye) [inline third_party/terraform/tests/resource_google_project_iam_policy_test.go]: moved to iam_test.go

**Comment by emilymye** [comment]
@danawillow added tests and moved a couple more things into utils

File: third_party/terraform/utils/iam_test.go [hunk 4]
```diff
@@ -0,0 +1,1113 @@
...
+		input  []*cloudresourcemanager.Binding
+		remove []*cloudresourcemanager.Binding
+		expect []*cloudresourcemanager.Binding
+	}{
+		{
+			input:  []*cloudresourcemanager.Binding{},
+			remove: []*cloudresourcemanager.Binding{},
+			expect: []*cloudresourcemanager.Binding{},
+		},
+		// Empty expect should no-op return empty
```
Comment (danawillow) [inline third_party/terraform/utils/iam_test.go]: nit: s/expect/input for most of these comments

File: third_party/terraform/utils/iam.go [hunk 5]
```diff
@@ -146,111 +151,186 @@ func iamPolicyReadModifyWrite(updater ResourceIamUpdater, modify iamPolicyModify
...
-			}
-			a.AuditLogConfigs = append(a.AuditLogConfigs, &alc)
+	am := createIamAuditConfigsMap(auditConfigs)
+	return listFromIamAuditConfigMap(am)
+}
+
+// Flattens AuditConfigs so each role has a single Binding with combined members\
+func removeAllAuditConfigsWithService(ac []*cloudresourcemanager.AuditConfig, service string) []*cloudresourcemanager.AuditConfig {
+	acMap := createIamAuditConfigsMap(ac)
+	if _, ok := acMap[service]; !ok {
```
[the remark this answers is not in the document]
Comment (emilymye) [reply third_party/terraform/utils/iam.go]: I deleted a log line but didn't realize - sorry!

File: third_party/terraform/utils/iam_test.go (hunk 4 above)
Comment (emilymye) [reply third_party/terraform/utils/iam_test.go]: intellij renaming is wild

**Pull request merged**